c_cpp_guide.md

```diff
@@ -0,0 +1,68 @@
+
+Quirk 3) In the current ewasm design, the Ethereum client writes data into WebAssembly as big-endian, and WebAssembly memory loads and stores are in little-endian. For example, when the call data is brought into memory using `callDataCopy`, and those bytes are loaded to the WebAssembly stack using `i64.load`, all of the bytes are reversed. So a function may be needed in C/C++ to reverse the bytes.
+
+Quirk 4) The output of compilers is a `.wasm` which may have some extra imports and exports with strange names. These must be manually fixed to be a valid ewasm contract.
+
+
+## Step-by-Step Guide
+
+The wasmception package offers patches to libc and libc++ which allow using `malloc` when targeting WebAssembly. This tool is not specific to ewasm, it is maintained by a third party. Warning: compiling wasmception will download llvm, llvm tools, musl C library, and libc++, which will require internet bandwidth, lots of RAM, and time.
+
+```sh 
+git clone https://github.com/yurydelendik/wasmception.git
+cd wasmception
+make -j4	# Warning: this required lots of bandwidth, RAM, and time. One hour on a mid-level laptop.
+cd ..
+``` 
+
+Write down the end of the output of the above make command, it should be something like: `--sysroot=/home/user/repos/wasmception/sysroot`.
+
+We modified the wasmception example into an ewasm contract, which we will now download. Make sure to edit the Makefile with the sysroot above, and change the path to our newly compiled version of clang to something like `/home/user/repos/wasmception/dist/bin/clang`. Make sure that `main.syms` has a list of ewasm helper functions you are using. If you are using C++, make sure to modify the Makefile to clang++, use `extern "C"` around the helper function declarations, and follow other tips from wasmception.
+
+```sh
+git clone https://gist.github.com/poemm/91b64ecd2ca2f1cb4a88d31315313b9b.git cwrc20
+cd cwrc20
+# edit the Makefile and main.syms as described above
+make
+```
+
+The output is `main.wasm`, but it needs some cleanup to remove some of it's extra import and export junk. For this, we use PyWebAssembly. 
+
+Aside: Alternatively, one can use a rust implementation of [wasm-chisel](https://github.com/wasmx/wasm-chisel), or install with `cargo install chisel`. But that requires a rust compiler.
+```
+cd ..
+git clone https://github.com/poemm/pywebassembly.git
+cd pywebassembly/examples/
+python3 ewasm_chisel.py ../../cwrc20/main.wasm
```

Comment:
Just checked, https://github.com/poemm/pywebassembly/blob/master/examples/ewasm_chisel.py doesn't behave like wasm-chisel (well, chisel is configurable through a config file, but talkinb about default behaviour).

chisel translates the form `env:ethereum_useGas` to `ethereum:useGas`, but only on recognised import names. ewasm_chisel translates any `env` namespace to `ethereum`. As a result it has a different naming requirement.

Reply:
Yes, `ewasm_chisel.py` is minimalist. It would be easy for `ewasm_chisel.py` to check against a list of eei method names, but I don't want to update it every time the eei changes. I have a similar problem with `main.syms`, which I instructed to user to check. It would be wise to just update a parsable list of current eei methods which can automate this.

My solution is to change the guide to reflect that they are different.

Reply:
The main difference is although the two tools are named similarly, they expect a different naming scheme from the user.

chisel expects the user to use functions like `ethereum_useGas` in C, while ewasm_chisel expects the user to use `useGas`.

Reply:
Good point. Changed `ewasm_chisel.py` to `ewasmify.py`.